README.md

# IP-Matching

[![GitHub package version](https://img.shields.io/github/v/release/SchoofsKelvin/ip-matching?include_prereleases&label=GitHub)](https://github.com/SchoofsKelvin/ip-matching)
[![NPM](https://img.shields.io/npm/v/ip-matching?label=NPM)](https://www.npmjs.com/package/ip-matching)
[![NPM downloads](https://img.shields.io/npm/dt/ip-matching?color=blue&label=NPM%20downloads)](https://www.npmjs.com/package/ip-matching)
[![Codacy Badge](https://api.codacy.com/project/badge/Grade/c62ef110c85246a58adc29ab8de2538d)](https://www.codacy.com/gh/SchoofsKelvin/ip-matching/dashboard)

[![GitHub Sponsors](https://img.shields.io/github/sponsors/SchoofsKelvin?color=green&label=GitHub%20Sponsors)](https://github.com/sponsors/SchoofsKelvin)
[![Donate](./.github/paypal.png)](https://www.paypal.me/KSchoofs)

Standalone module with some handy features regarding IPs:
- Quick and easy-to-use way to check whether an IP meets a requirement
- Support for IPv4/IPv6 wildcard addresses, ranges, subnetworks and masks.
- Supports both parsing and outputting all common IPv6 notations
- Utility methods, e.g. next/previous IP, range to a list of CIDRs, ...
- Every release is thoroughly [tested](https://github.com/SchoofsKelvin/ip-matching/tree/master/src/) and [linted](https://github.com/SchoofsKelvin/ip-matching/actions) before published

## Installation
```bash
npm install --save ip-matching
# or
yarn add ip-matching
```

Comes with its own built-in [TypeScript declarations](https://github.com/SchoofsKelvin/ip-matching/blob/master/api/index.d.ts)
 with included documentation.

## Example
```ts
import { getMatch, IPMatch, IPSubnetwork, IPRange, matches } from 'ip-matching';

// matches(ip: string | IP, target: string | IPMatch): boolean;

matches('10.0.0.1', '10.0.0.0/24'); // true
matches('10.0.1.1', '10.0.0.0/24'); // false
matches('abc::def', 'abc:*::def'); // true
matches('abc::def', 'abc:9::def'); // false
matches('0001:2:3:4:5:6:7', '1:2:3:4:5:6:7'); // true

// getMatch returns an instance of
// IPv4, IPv6, IPRange, IPSubnetwork or IPMask, all extending IPMatch
const mySubnet: IPMatch = getMatch('fefe::0001:abcd/112');
mySubnet.type; // 'IPSubnetwork'
mySubnet instanceof IPSubnetwork; // true
mySubnet instanceof IPMatch; // true
mySubnet.toString(); // 'fefe::1:0/112'
mySubnet.matches('FEFE::1:bbbb'); // true
mySubnet.matches('FEFE::2:bbbb'); // false
mySubnet.equals(new IPSubnetwork(new IPv6('fefe::1:abcd'), 112)); // true
mySubnet.getAmount(); // 65536
(mySubnet as IPSubnetwork).getLast().toString(); // 'fefe::1:ffff'

const myIp = new IPv6('a:0:0::B:0:C');
myIp.toString(); // 'a::b:0:c'
myIp.toLongString(); // 'a:0:0:0:0:b:0:c'
myIp.toFullString(); // '000a:0000:0000:0000:0000:000b:0000:000c'
new IPv6('::ffff:a9db:*').toMixedString(); // '::ffff:169.219.*.*'

const myRange = getMatch('10.0.0.0-10.1.2.3') as IPRange;
myRange.convertToMasks().map((mask: IPMask) => mask.convertToSubnet().toString());
// [ '10.0.0.0/16', '10.1.0.0/23', '10.1.2.0/30' ]

const mask1 = getMatch('10.0.1.0/255.0.255.0') as IPMask;
const mask2 = getMatch('10.0.0.0/128.0.0.0') as IPMask;
mask1.isSubsetOf(mask2); // true
mask2.getAmount(); // 2147483648

getMatch('a::abbc:1234/ffff::ff80:000f').toString(); // 'a::ab80:4/ffff::ff80:f'
```
***Note**: The `matches` function and all constructors error for invalid inputs*

You can take a look at the [test code](https://github.com/SchoofsKelvin/ip-matching/blob/master/src/ip.test.ts) or the [TypeScript declarations](https://github.com/SchoofsKelvin/ip-matching/blob/master/api/index.d.ts) for all the features.

## Allowed patterns
* IP (IPv4/IPv6)
    * Regular IPv4: `10.0.0.0`
    * Wildcard IPv4: `10.0.0.*` or even `10.*.0.*`
    * Regular IPv6: `2001:0db8:85a3:0000:0000:8a2e:0370:7334`
    * Shortened IPv6: `2001:db8:85a3::8a2e:0370:7334` or `::` or `::1` or `a::`
    * Wildcard IPv6: `2001::*` or even `2001::*:abc:*`
    * Mixed IPv6 (mapped IPv4): `::ffff:127.0.0.1` _(no wildcards allowed in IPv4 part)_
    * **Not allowed**: `10.0.1*.0` or `2001::a*c`
* IP Range
    * IPv4: `10.0.0.0-10.1.2.3`
    * IPv6: `2001::abc-2001::1:ffff`
    * **Note**: Left side has to be "lower" than the right side
* IP Subnetwork
    * IPv4: `10.0.0.0/16`
    * IPv6: `2001::/123`
* IP Mask
    * IPv4: `10.0.0.0/255.0.64.0`
    * IPv6: `2001:abcd::/ffff:ff8::`

The IPMask's `toString()` method does not automatically simplify e.g. `/255.0.0.0` to `/8`, even though those are equivalent. Since 2.0.0, IPMask comes with a method `convertToSubnet()` which returns an equivalent IPSubnetwork if possible, otherwise `undefined`.

## Links
* GitHub: https://github.com/SchoofsKelvin/ip-matching
* NPM: https://www.npmjs.com/package/ip-matching


CHANGELOG.md

# Changelog

Keep track of all changes between versions. Should basically match the release descriptions on GitHub.

## 2.1.2

### Project changes
- Updated description (in `package.json`)
  - This module is more like 50% matching and 50% IP-related utilities
- Updated README
  - Added badges for GitHub/NPM and Codacy grade
  - Reworded certain parts and improved formatting
  - Improved the example, showcasing more functionality
- Added CHANGELOG

### Development changes
- Renamed script names to be more logical/consistent
- Add Codacy for static analysis of source code

## 2.1.1

### Development changes
- Added ESLint
  - Setup for TypeScript code base (including tests)
  - Includes VS Code build task in watch mode with problem matcher
  - Included in the `prepack` script, thus running in CI for every release and pushed commit
- Typescript updated to `^4.3.5`
- `@microsoft/api-extractor` updated to `^7.18.1`
- Small refactor of unreleased code

## 2.1.0

### New features
- Added `toBits()` to `IPv4`/`IPv6`, producing e.g. `[1, 1, 0, 0, 0, 0, ...]` for `192.0.0.0` (length 32 and 128 for IPv4 and IPv6 respectively)
- Added `IPv4.fromBits(bits)` and `IPv6.fromBits(bits)`, accepting the "bit arrays" produced from `toBits()`
  - Accepts arrays (with length 32 for `IPv4` and length 128 for `IPv6`) filled with `0`'s and `1`'s
- Added `IPv4.bits = 32` and `IPv6.bits = 128`, both on the class and class objects (e.g. `new IPv4('0.0.0.0').bits`)
- Added `getAmount()` to `IPMatch`, allowing you see how many IP addresses a match (e.g. a range/subnet) matches
- Added `isSubsetOf(mask)` to `IPMask`, allowing to check whether one mask is a subset of another mask
  - Mask A is a subset of mask B if every IP address that is matched by A is also matched by B

### Changes
- Caching added to `convertToMasks()`/`convertToSubnets()`, making repeated calls a lot cheaper
- `matches(ip)` in `IPRange` no longer errors for passing an `IPv4` to an `IPv6` range and vice-versa, instead returning false now
  - This is line with `matches()` in general, where it only errors if the input is neither an IP nor a string that can be converted to one


## 2.0.0
Lots of changes since the previous v1.2.1 version.

### New features
- Added `getNext()` and `getPrevious()` to `IPv4` and `IPv4`
- Added `getFirst()` and `getLast()` to `IPRange` and `IPSubnetwork`
- Added `IPMask` subclass, supporting _any_ mask _(e.g. `255.0.128.0`)_
- Added `convertToSubnets()` to `IPRange` (optimized CIDR array)
- Added `convertToSubnet()` to `IPMask` _(only supports traditional subnet masks)_
- Added `convertToMasks()` to `IPMatch`

### Improved documentation
- Improved JSDoc comments in the source code
- API Extractor generating an index.d.ts for `package.json` `"typings"`
- Added workflow file (GitHub Actions)
  - Tries to compile the code
  - Runs API Extractor (in validation mode)
  - Runs all the tests
  - Packages an NPM-ready tarball (uploaded as artifact)
  - Prepares a release draft for tags starting with `v`
  
### Bug fixes
- Improved `IPSubnetwork` parsing _(supports `/0` now)_

### Development changes
- Added API Extractor support
- Switched to Jest for testing
- Removed unused files (e.g. ESLint)
- Updated .gitignore, .vscode tasks/settings, ...
- License change to LGPL-3.0-only


src/test.ts

require('source-map-support/register');

import './ip.test';



src/utils.test.ts

import { getMatch, IP, IPMask, IPMatch, IPv4, IPv6, partsToIP } from './ip';
import { compactMasks } from './utils';

const toString = (obj: any): string => `${obj}`;

function validateMatches(original: IPMatch[], processed: IPMatch[], ips: IP[]) {
  for (const ip of ips) {
    const a = original.some(m => m.matches(ip));
    const b = processed.some(m => m.matches(ip));
    if (a === b) continue;
    fail(`Expected ${a} for ${ip} but got ${b}`);
  }
}
function validateRandom(original: IPMatch[], processed: IPMatch[], clazz: typeof IPv4 | typeof IPv6) {
  const ips: IP[] = [];
  for (let i = 0; i < 1000; i++) {
    const bits: number[] = [];
    for (let i = 0; i < clazz.bits; i++) bits[i] = Math.random() < 0.5 ? 0 : 1;
    ips.push(clazz.fromBits(bits));
  }
  validateMatches(original, processed, ips);
}
function findEdgeCases(matches: IPMatch[]) {
  const cases: IP[] = [];
  for (const match of matches) {
    for (const mask of match.convertToMasks()) {
      cases.push(mask.ip);
      const prev = mask.ip.getPrevious();
      if (prev) cases.push(prev);
      const next = mask.ip.getNext();
      if (next) cases.push(next);
      const ipParts = [...mask.ip.parts];
      const maskParts = mask.mask.parts;
      const maxPart = (2 ** (mask.ip.bits / ipParts.length)) - 1;
      for (let i = 0; i < ipParts.length; i++) {
        ipParts[i] = ipParts[i] | (maxPart & maskParts[i]);
      }
      const ip = partsToIP(ipParts);
      cases.push(ip);
      const ipPrev = ip.getPrevious();
      if (ipPrev) cases.push(ipPrev);
      const ipNext = ip.getNext();
      if (ipNext) cases.push(ipNext);
    }
  }
  return cases;
}
function validateEdges(original: IPMatch[], processed: IPMatch[]) {
  const ips = [...findEdgeCases(original), ...findEdgeCases(processed)];
  validateMatches(original, processed, ips);
}

describe('compactMasks', () => {
  describe('IPv4 set 1', () => {
    const masks = [
      getMatch('10.0.0.0/255.0.255.0') as IPMask,
      getMatch('10.0.0.0/255.0.255.0') as IPMask,
      getMatch('10.0.0.0/255.255.1.0') as IPMask,
      getMatch('10.0.0.0/255.255.0.255') as IPMask,
      getMatch('11.0.0.0/255.0.255.0') as IPMask,
      getMatch('20.1.2.3/255.0.0.0') as IPMask,
      getMatch('20.1.2.3/0.255.0.0') as IPMask,
      getMatch('20.1.2.3/255.255.0.0') as IPMask,
    ];
    const compacted = compactMasks(masks);
    const expected = [
      getMatch('0.1.0.0/0.255.0.0') as IPMask,
      getMatch('10.0.0.0/254.0.255.0') as IPMask,
      getMatch('10.0.0.0/255.255.0.255') as IPMask,
      getMatch('10.0.0.0/255.255.1.0') as IPMask,
      getMatch('20.0.0.0/255.0.0.0') as IPMask,
    ].map(toString).sort();
    test('matches expected', () => {
      expect(compacted.map(toString).sort()).toEqual(expected);
    });
    test('validate edge cases', () => {
      validateEdges(masks, compacted);
    });
    test('validate 1000 random addresses', () => {
      validateRandom(masks, compacted, IPv4);
    })
    test('validate first two octets', () => {
      let ip = getMatch('0.0.0.0') as IPv4;
      const parts = ip.parts;
      for (let i = 0; i <= 255 * 255; i++) {
        parts[0] = i >> 8;
        parts[1] = i & 255;
        ip = partsToIP(parts) as IPv4;
        const a = masks.some(m => m.matches(ip));
        const b = compacted.some(m => m.matches(ip));
        if (a === b) continue;
        fail(`Expected ${a} for ${ip} but got ${b}`);
      }
    });
  });

  describe('IPv6 set 1', () => {
    const masks = [
      getMatch('a:b:c:d::/ffff::') as IPMask,
      getMatch('a:b:c:d::/ffff::') as IPMask,
      getMatch('a:b:c:d::/0:ffff::') as IPMask,
      getMatch('a:b:c:d::/ffff:ffff::') as IPMask,
      getMatch('b:b:c:d::/ffff::') as IPMask,
      getMatch('fa:b:c:d::/ffff::') as IPMask,
      getMatch('fb:b:c:d::/ffff::') as IPMask,
      getMatch('8888:1:2:3::/ffff::') as IPMask,
      getMatch('8888:1:2:3::/0:ffff::') as IPMask,
      getMatch('8888:1:2:3::/ffff:ffff::') as IPMask,
    ];
    const compacted = compactMasks(masks);
    const expected = [
      getMatch('0:1::/0:ffff::') as IPMask,
      getMatch('0:b::/0:ffff::') as IPMask,
      getMatch('a::/fffe::') as IPMask,
      getMatch('fa::/fffe::') as IPMask,
      getMatch('8888::/ffff::') as IPMask,
    ].map(toString).sort();
    test('matches expected', () => {
      expect(compacted.map(toString).sort()).toEqual(expected);
    });
    test('validate edge cases', () => {
      validateEdges(masks, compacted);
    });
    test('validate 1000 random addresses', () => {
      validateRandom(masks, compacted, IPv6);
    })
    test('validate first hextet', () => {
      let ip = getMatch('0::') as IPv6;
      const parts = ip.parts;
      for (let i = 0; i <= 0xffff; i++) {
        parts[0] = i;
        ip = partsToIP(parts) as IPv6;
        const a = masks.some(m => m.matches(ip));
        const b = compacted.some(m => m.matches(ip));
        if (a === b) continue;
        fail(`Expected ${a} for ${ip} but got ${b}`);
      }
    });
  });
});



hotreload.js

const req = require;
const fs = require('fs');

function hotreloadModule(id) {
    id = req.resolve(id);
    const mod = req.cache[id];
    if (!mod) return false;
    delete req.cache[id];
    const exp = req(id);
    // Object.assign(mod.exports, exp);
    // mod.exports = exp;
    console.log('Hotreloaded', id);
}

fs.watch(__dirname, { recursive: true }, (event, filename) => {
    if (filename === 'hotreload.js') return;
    try {
        hotreloadModule('./' + filename);
    } catch (e) {
        // console.error('Hotreload failed:', e);
    }
});

const ModuleProxy = {
    get(mod, key) {
        mod = require.cache[mod.id] || mod;
        return mod.real_exports[key];
    },
    ownKeys(mod) {
        mod = require.cache[mod.id] || mod;
        return Object.keys(mod.real_exports);
    },
    has(mod, key) {
        mod = require.cache[mod.id] || mod;
        return key in (mod.real_exports)[key];
    },
    getPrototypeOf(mod) {
        return Object.getPrototypeOf((require.cache[mod.id] || mod).real_exports);
    },
    getOwnPropertyDescriptor(mod, key) {
        return Object.getOwnPropertyDescriptor((require.cache[mod.id] || mod).real_exports, key);
    }
}

function wrapExtension(ext) {
    const old = require.extensions[ext];
    require.extensions[ext] = (mod, file) => {
        const res = old(mod, file);
        mod.real_exports = mod.exports;
        mod.exports = new Proxy(mod, ModuleProxy);
        return mod;
    }
}

wrapExtension('.js');
wrapExtension('.ts');

src/collection.ts
import { IPMask, IPv4, IPv6 } from './ip';

type Compacter = (a: IPMask, b: IPMask) => IPMask | undefined;
const COMPACTERS: Compacter[] = [
    // Simple compacter for when one of the subset of the other
    (a, b) => a.isSubsetOf(b) ? b : (b.isSubsetOf(a) ? a : undefined),
    // Compact if masks are same but IPs are different
    (a, b) => {
        if (!a.mask.equals(b.mask)) return undefined;
        const iBitsA = a.ip.toBits();
        const iBitsB = b.ip.toBits();
        const mBits = a.mask.toBits();
        let changed = false;
        for (let i = 0; i < iBitsA.length; i++) {
            if (!mBits[i]) continue; // mask=0 so bit can be both already
            if (iBitsA[i] === iBitsB[i]) continue; // bits are same anyway
            if (changed) return undefined; // only works if the IPs differ in 1 bit
            mBits[i] = 0; // flag the bit as being able to be both
            changed = true;
        }
        const ipType = a.ip.type === 'IPv4' ? IPv4 : IPv6;
        const mask = new IPMask(a.ip, ipType.fromBits(mBits));
        // Wouldn't this mean the IPs are the same? First compacter should've caught that?
        //if (mask.equals(a) || mask.equals(b)) return undefined;
        return mask;
    },
];

export function compactMasks(masks: IPMask[]): IPMask[] {
    masks = [...masks];
    if (masks.length < 2) return masks;
    let results: IPMask[] = [];
    masks: for (const mask of masks) {
        for (let i = 0; i < results.length; i++) {
            const other = results[i];
            for (const compacter of COMPACTERS) {
                const compacted = compacter(mask, other);
                if (!compacted) continue;
                // Remove whatever we compacted with from the results
                results = [...results.slice(0, i), ...results.slice(i + 1)];
                masks.push(compacted);
                continue masks;
            }
        }
        // Did not compact with a previous result
        results.push(mask);
    }
    return results;
}


api/untrimmed.d.ts
/**
 * Library mostly focused on IP matching, but with some other fancy goods.
 * Repository: https://github.com/SchoofsKelvin/ip-matching
 * @packageDocumentation
 */

/**
 * Tries to convert the given input string to an IP, aka an IPv4 or IPv6 object.
 * For ease-of-use, if the input is already an IPv4 or IPv6, it is returned.
 * @throws Errors if the given input format matches an IPv4/IPv6 address well enough, but is still invalid.
 */
export declare function getIP(input: string | IP): IP | null;

/**
 * Converts a string to an IPMatch object. This correspondends either to
 * an IPv4, IPv4, IPRange or IPSubnetwork object, all extending the IPMatch class.
 * For ease-of-use, if the given input is an IPMatch object, that object itself is returned.
 * @param input - The input string to convert, or IPMatch object to return.
 * @returns Returns an IPMatch for the given string (or returns the given IPMatch itself)
 */
export declare function getMatch(input: string | IPMatch): IPMatch;

/** Represents either an IPv4 or an IPv6, aka single addresses (or wildcard ones) */
export declare type IP = IPv4 | IPv6;

/** Represents an IP mask. The combination of an IP and a mask. A more complex version of IPSubnetwork. */
export declare class IPMask extends IPMatch {
    readonly ip: IP;
    readonly mask: IP;
    readonly type = "IPMask";
    readonly input: string;
    constructor(ip: IP, mask: IP);
    /** Checks whether the given IP matches this mask */
    matches(ip: string | IP): boolean;
    equals(match: IPMatch): boolean;
    /**
     * Converts this IPMask to a string, by joining the IP and mask with a slash, e.g. "IP/mask".
     * Does simplify the IP and mask in their IP form, but does not simplify e.g. `10.0.0.0/255.0.0.0` to `10.0.0.0/8`.
     */
    toString(): string;
    /** @internal */
    private static convertToSubnet;
    /**
     * Tries to convert this IPMask to an IPSubnetwork. This only works if this mask is a "proper" subnet mask.
     * In other words, the bits have to be sequential. `255.255.128.0` is valid, `255.255.63.0` is not.
     * When this is not the case, `undefined` is returned instead.
     */
    convertToSubnet(): IPSubnetwork | undefined;
    convertToMasks(): IPMask[];
    getAmount(): number;
    /**
     * Returns whether this mask is a subset of the given mask. In other words, all IP addresses matched
     * by this mask should also be matched by the given mask, although the given mask can match others too.
     * @throws Throws an error if the IP address types mismatch (e.g. this mask is for IPv4 but the given is IPv6)
     */
    isSubsetOf(mask: IPMask): boolean;
}

/**
 * Superclass of the IPv4, IPv6, IPRange and IPSubnetwork classes.
 * Only specifies a generic .matches() function and .type field.
 *
 * **Check the specific classes for more specialized methods/docs**
 * e.g. IPRange comes with `convertToSubnets`, IPv6 with `toLongString`, ...
 */
export declare abstract class IPMatch {
    /** String indicating the type of this IPMatch, as an alternative to `instanceof`. Check subclasses for the possible values */
    abstract readonly type: string;
    /** The string representation of this IPMatch. Not necessarily the exact input string that generated it */
    abstract readonly input: string;
    /**
     * This used to be the generic way of converting a string to an IPRange/IPv4/... without assuming a type.
     * This class is now made abstract with a protected constructor, in favor of the new `getMatch(input)` function.
     * The abstract/deprecated/protected flag are to warn users about switching over to the new function.
     * With the way TypeScript compiles them to JavaScript, this constructor still works (thus compatible with old code)
     * @deprecated Use `getMatch(input: string)` instead.
     */
    protected constructor(input: string | null);
    /**
     * Checks whether the given IP (or string to be first converted to an IP) matches this IPMatch object.
     * The exact way this is checked is specific to each IPv4/IPRange/... class. Check their documentation.
     */
    abstract matches(ip: string | IP): boolean;
    /** Each subclass formats itself in a specific way. IPv6 also has a bunch of extra string methods. Check their documentation */
    abstract toString(): string;
    /**
     * Checks whether this IPMatch equals the given match. The match type matters, e.g. the IPv4 `10.0.0.0` and
     * the IPSubnetwork `10.0.0.0/32` will result in this method returning false, even though they both only match `10.0.0.0`.
     */
    abstract equals(match: IPMatch): boolean;
    /**
     * Converts this IPMatch to a list of IPMasks (union) matching the same IPs.
     * IPRange has a handy method convertToSubnets() to convert the range to an array
     * of IPSubnetworks, which are basically CIDR notations. If you're looking at
     * this method, you might also be interested in checking `convertToSubnets` out.
     */
    abstract convertToMasks(): IPMask[];
    /** Retuns the amount of unique IP addresses this IPMatch would match */
    abstract getAmount(): number;
}

/** Represents a range of IP addresses, according to their numerical value */
export declare class IPRange extends IPMatch {
    readonly left: IP;
    readonly right: IP;
    readonly type = "IPRange";
    readonly input: string;
    /** Both values should be the same type (IPv4 or IPv6) and `left` should be lower in numeric value than `right` */
    constructor(left: IP, right: IP);
    /** Checks whether the given IP lies in the range defined by the two bounds (inclusive) */
    matches(ip: string | IP): boolean;
    equals(match: IPMatch): boolean;
    /** Converts this IPRange to a string, by joining the two bounds with a dash, e.g. "IP1-IP2" */
    toString(): string;
    /** @internal */
    private static convertToSubnets;
    /** Converts this IPRange to an optimized list of (CIDR) IPSubnetworks */
    convertToSubnets(): IPSubnetwork[];
    /** @internal */
    private static convertToMasks;
    convertToMasks(): IPMask[];
    getAmount(): number;
    /** Returns the first IP address in this range */
    getFirst(): IP;
    /** Returns the last IP address in this range */
    getLast(): IP;
    protected isLowerOrEqual(left: IP, right: IP): boolean;
}

/** Represents a subnetwork. The combination of an IP and a (simple) mask. A simplified version of IPMask. */
export declare class IPSubnetwork extends IPMatch {
    readonly bits: number;
    readonly type = "IPSubnetwork";
    readonly input: string;
    protected range: IPRange;
    /** Bits has to be in the range 0-32 for IPv4 and 0-128 for IPv6 */
    constructor(ip: IP, bits: number);
    /** Checks whether the given IP lies in this subnetwork */
    matches(ip: string | IP): boolean;
    equals(match: IPMatch): boolean;
    /** Converts this IPSubnetwork to a string in CIDR representation, e.g. "IP/mask" */
    toString(): string;
    /** @internal */
    private static convertToMasks;
    convertToMasks(): IPMask[];
    getAmount(): number;
    /** Returns the first IP address in this range */
    getFirst(): IP;
    /** Returns the last IP address in this range */
    getLast(): IP;
}

/** Represents an IPv4 address, optionall with wildcards */
export declare class IPv4 extends IPMatch {
    readonly type = "IPv4";
    readonly parts: number[];
    readonly input: string;
    constructor(input: string);
    /**
     * Checks whether the given IP (or string to be first converted to an IP) matches this IPv4 object.
     * - If the given string represents an IPv6 address, this method returns false.
     * - In other cases, for an IPv4, we check if all 4 octets match.
     * - Octets that are wildcards in this object are always assumed to match.
     * - Octets that are wildcards in the input are **NOT** seen as a wildcard, e.g.
     *    `10.0.0.*` matches `10.0.0.3`, but the inverse would give false.
     */
    matches(ip: string | IP): boolean;
    equals(match: IPMatch): boolean;
    /** Returns whether this IPv4 is exact (aka contains no wildcards) */
    exact(): boolean;
    /**
     * Returns this IPv4 in dot-decimal/quat-dotted notation. Wildcards are represented as stars.
     * For example: `"10.*.0.*"`
     */
    toString(): string;
    /** @internal */
    private static convertToMasks;
    convertToMasks(): IPMask[];
    getAmount(): number;
    /**
     * Returns the previous address, or undefined for `0.0.0.0`.
     * In case of a non-exact IP, the wildcard parts are ignored.
     * E.g. getPrevious for `10.0.*.0` returns `9.255.*.255`
     */
    getPrevious(): IPv4 | undefined;
    /**
     * Returns the next address, or undefined for `255.255.255.255`.
     * In case of a non-exact IP, the wildcard parts are ignored.
     * E.g. getNext for `10.0.*.255` returns `10.1.*.0`
     */
    getNext(): IPv4 | undefined;
    /** Converts this IP to an array of bits, e.g. `[1, 1, 0, 0, 0, ...]` for `192.0.0.0`. */
    toBits(): number[];
    /** Converts an array of 32 bits to an IPv4, e.g. `192.0.0.0` for `[1, 1, 0, 0, 0, ...]` */
    static fromBits(bits: number[]): IPv4;
    /** Field present on both IPv4 and IPv6 indicating how many bits an address of that type has */
    static readonly bits = 32;
    /** Field present on both IPv4 and IPv6 addresses indicating how many bits an address of that type has */
    readonly bits = 32;
}

/** Represents an IPv6 address, optionall with wildcards */
export declare class IPv6 extends IPMatch {
    readonly type = "IPv6";
    readonly parts: number[];
    readonly input: string;
    constructor(input: string);
    /**
     * Checks whether the given IP (or string to be first converted to an IP) matches this IPv6 object.
     * - If the given string represents an IPv4 address, this method returns false.
     * - In other cases, for an IPv6, we check if all 8 hextets/hexadectets match.
     * - Octets that are wildcards in this object are always assumed to match.
     * - Octets that are wildcards in the input are **NOT** seen as a wildcard, e.g.
     *    `2001::abcd:*` matches `2001::abcd:1`, but the inverse would give false.
     */
    matches(ip: string | IP): boolean;
    equals(match: IPMatch): boolean;
    /** Returns whether this IPv4 is exact (aka contains no wildcards) */
    exact(): boolean;
    /** Returns an array with the 8 hextets of this address, or `"*"` for wildcard hextets */
    toHextets(): string[];
    /**
     * Returns the address in the full format, but with leading zeroes of hextets omitted.
     * Hextets representing wildcards will be shown as `"*"` instead.
     * Example result: `"2001:0:0:0:0:0:abc:1"`
     */
    toLongString(): string;
    /**
     * Returns the address in the full format, but without omitting leading zeroes or hextets.
     * Hextets representing wildcards will be shown as `"*"` instead.
     * Example result: `"2001:0000:0000:0000:0000:0000:0abc:0001"`
     */
    toFullString(): string;
    /** Returns a mixed address (32 last bits representing an IPv4 address) in a mixed format e.g. "::ffff:c000:0280" as "::ffff:192.0.2.128" */
    toMixedString(): string;
    /**
     * Returns the address in the shortest possible format, according to RFC 5952:
     * - All hexadecimal digits are lowercase (if applicable), as is the case with .toLongString(), toFullString(), ...
     * - Leading zeroes of each hextet are suppressed, apart from the all-zero field which is rendered as a single zero
     * - The (leftmost) longest sequence of multiple consecutive all-zero hextets is replaced with "::"
     * - If this address is known to be IPv4 mapped, it is displayed as such, which currently are for e.g. 127.0.0.1:
     *    - `"::ffff:127.0.0.1"`
     *    - `"::ffff:0:127.0.0.1"`
     */
    toString(): string;
    /** @internal */
    private static convertToMasks;
    convertToMasks(): IPMask[];
    getAmount(): number;
    /**
     * Returns the previous address, or undefined for `::`.
     * In case of a non-exact IP, the wildcard parts are ignored.
     * E.g. getPrevious for `::5:*:0` returns `::4:*:ffff`
     */
    getPrevious(): IPv6 | undefined;
    /**
     * Returns the next address, or undefined for `ffff:ffff:ffff:ffff:ffff:ffff:ffff:ffff`.
     * In case of a non-exact IP, the wildcard parts are ignored.
     * E.g. getNext for `::0:*:ffff` returns `::1:*:0`
     */
    getNext(): IPv6 | undefined;
    /** Converts this IP to an array of bits, e.g. `[1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0, ...]` for `f8::`. */
    toBits(): number[];
    /** Converts an array of 128 bits to an IPv6, e.g. `f8::` for `[1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0, ...]` */
    static fromBits(bits: number[]): IPv6;
    /** Field present on both IPv4 and IPv6 indicating how many bits an address of that type has */
    static readonly bits = 128;
    /** Field present on both IPv4 and IPv6 addresses indicating how many bits an address of that type has */
    readonly bits = 128;
}

/**
 * Checks whether the given IP matches the given whitelist
 * @param ip - The IP to check, or a string to be converted to an exact IP
 * @param target - The target to check against, or a string to convert to an IPMatch
 * @throws Throws an error if either argument is a string but does not have a correct format
 */
export declare function matches(ip: string | IP, target: string | IPMatch): boolean;

export { }
